### PR TITLE
coord,sql: support CREATE TEMPORARY TABLE

### PR DESCRIPTION
**Pull request by JLDLaughlin**
This change, along with https://github.com/MaterializeInc/materialize/pull/5542, fixes https://github.com/MaterializeInc/materialize/issues/5470. 

This change allows users to create temporary tables in their
Materialize instance. Temporary tables are only visible to the
user that created them, and are dropped when their connection is
dropped or the instance is stopped. Temporary tables behave identically
to other temporary objects.

The new `CREATE TABLE` grammar looks like:
Screen Shot 2021-02-01 at 4 11 13 PM

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5546)

### CODE REVIEW

**Comment by philip-stoev** [comment]
@JLDLaughlin do you mind if I join the review? There are some tests I could add

**Comment by JLDLaughlin** [comment]
@philip-stoev - no, that would be great! thank you!

**Review by benesch — APPROVED** [review]
LGTM—but pending Philip's stuff!

File: test/testdrive/temporary.td
Comment (philip-stoev) [inline test/testdrive/temporary.td:70]: I am afraid GitHub does not allow global-yet-individually-resolveable conversations in the general Comments section, so I will try adding my stuff as individual comments against the test file instead.

Item #1. Temporary tables do not behave correctly if there is an actual table with the same name. The temporary table needs to mask the existing one while it is alive.

In this branch:
```
pstoev=> create table t1 (f1 integer);
CREATE TABLE
pstoev=> insert into t1 values (1);
INSERT 0 1
pstoev=> create temporary table t1 (f1 integer);
CREATE TABLE
pstoev=> insert into t1 values (2);
INSERT 0 1
pstoev=> select * from t1;
 f1 
----
  1
  2 <- wrong, the two inserts got mixed together at some point
(2 rows)

pstoev=> drop table t1;
DROP TABLE
pstoev=> select * from t1;
 f1 
----
(0 rows) <- wrong, the data from the non-temporary table should have shown up.
```

Contrast with Postgresql:

```
philips=> create table t1 (f1 integer);
CREATE TABLE
philips=> insert into t1 values (1);
INSERT 0 1
philips=> create temporary table t1 (f1 integer);
CREATE TABLE
philips=> insert into t1 values (2);
INSERT 0 1
philips=> select * from t1;
 f1 
----
  2 <- data from the temporary table is visible here
(1 row)

philips=> drop table t1;
DROP TABLE
philips=> select * from t1;
 f1 
----
  1 <- data from the original table is returned
```

File: test/testdrive/temporary.td
Comment (philip-stoev) [inline test/testdrive/temporary.td:71]: the ```/dt``` command in ```psql``` does not list the temporary tables. Contrast with Postgres:

```
philips=> create temporary table t1 (f1 integer);
CREATE TABLE
philips=> \dt
         List of relations
  Schema   | Name | Type  |  Owner  
-----------+------+-------+---------
 pg_temp_3 | t1   | table | philips
(1 row)
```

File: test/testdrive/temporary.td
Comment (philip-stoev) [inline test/testdrive/temporary.td:72]: The error message when mixing temporary and non-temporary objects references an internal identifier rather than the actual name of the thing. But I guess this is a more general problem not limited to temporary tables alone:

```
pstoev=> create temporary table t1 (f1 integer);
CREATE TABLE
pstoev=> create view v1 as select (select * from t1 limit 1) from t1;
ERROR:  non-temporary items cannot depend on temporary item 'u47'
```

In postgres, the view is automatically converted to a temporary one.

File: test/testdrive/temporary.td
Comment (philip-stoev) [inline test/testdrive/temporary.td:74]: In the presence of a non-temporary and a temporary table with the same name, ```DROP TABLE``` drops the non-temporary table:

```
pstoev=> create table t1 (f1 integer);
CREATE TABLE
pstoev=> create temporary table t1 (f1 integer);
CREATE TABLE
pstoev=> drop table t1; <- this seems to drop the non-temporary table
DROP TABLE
pstoev=> explain select * from t1;
      Optimized Plan      
--------------------------
 %0 =                    +
 | Get mz_temp.t1 (u2079)+ <- the temporary table remained
(1 row)
```

File: test/testdrive/temporary.td
Comment (philip-stoev) [inline test/testdrive/temporary.td:76]: ```ALTER TABLE RENAME TO``` causes an index out of bounds panic

```
pstoev=> create temporary table t1 (f1 integer);
CREATE TABLE
pstoev=> alter table t1 rename to t2;
server closed the connection unexpectedly

2021-02-03T08:38:37.077042Z ERROR panic: <unnamed>: index out of bounds: the len is 2 but the index is 2
   0: materialized::handle_panic
             at src/materialized/src/bin/materialized/main.rs:569:65
   1: core::ops::function::Fn::call
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/ops/function.rs:70:5
   2: std::panicking::rust_panic_with_hook
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:581:17
   3: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:484:9
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys_common/backtrace.rs:153:18
   5: rust_begin_unwind
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:483:5
   6: core::panicking::panic_fmt
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/panicking.rs:85:14
   7: core::panicking::panic_bounds_check
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/panicking.rs:62:5
   8: <usize as core::slice::index::SliceIndex<[T]>>::index_mut
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/slice/index.rs:261:14
      core::slice::index::<impl core::ops::index::IndexMut<I> for [T]>::index_mut
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/slice/index.rs:26:9
      <alloc::vec::Vec<T> as core::ops::index::IndexMut<I>>::index_mut
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/vec.rs:1994:9
      sql::ast::transform::create_stmt_rename
             at src/sql/src/ast/transform.rs:35:13
   9: coord::catalog::CatalogItem::rename_item_refs::{{closure}}
             at src/coord/src/catalog.rs:329:17
  10: coord::catalog::CatalogItem::rename_item_refs
             at src/coord/src/catalog.rs:339:32
  11: coord::catalog::Catalog::transact
             at src/coord/src/catalog.rs:1285:32
  12: coord::coord::Coordinator<C>::catalog_transact::{{closure}}
             at src/coord/src/coord.rs:2789:22
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/future/mod.rs:80:19
  13: coord::coord::Coordinator<C>::sequence_alter_item_rename::{{closure}}
             at src/coord/src/coord.rs:2752:15
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/future/mod.rs:80:19
      coord::coord::Coordinator<C>::sequence_plan::{{closure}}
             at src/coord/src/coord.rs:1660:17
  14: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/future/mod.rs:80:19
      coord::coord::Coordinator<C>::message_statement_ready::{{closure}}
             at src/coord/src/coord.rs:500:17
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/future/mod.rs:80:19
      coord::coord::Coordinator<C>::serve::{{closure}}
             at src/coord/src/coord.rs:402:21
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/future/mod.rs:80:19
  15: tokio::park::thread::CachedParkThread::block_on::{{closure}}
             at /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.1.0/src/park/thread.rs:263:54
      tokio::coop::with_budget::{{closure}}
             at /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.1.0/src/coop.rs:121:9
      std::thread::local::LocalKey<T>::try_with
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/thread/local.rs:272:16
      std::thread::local::LocalKey<T>::with
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/thread/local.rs:248:9
      tokio::coop::with_budget
             at /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.1.0/src/coop.rs:114:5
      tokio::coop::budget
             at /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.1.0/src/coop.rs:98:5
      tokio::park::thread::CachedParkThread::block_on
             at /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.1.0/src/park/thread.rs:263:31
  16: tokio::runtime::enter::Enter::block_on
             at /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.1.0/src/runtime/enter.rs:151:13
      tokio::runtime::thread_pool::ThreadPool::block_on
             at /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.1.0/src/runtime/thread_pool/mod.rs:71:9
  17: tokio::runtime::Runtime::block_on
             at /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.1.0/src/runtime/mod.rs:452:43
  18: coord::coord::serve::{{closure}}::{{closure}}
             at src/coord/src/coord.rs:3421:31
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys_common/backtrace.rs:137:18
  19: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/thread/mod.rs:464:17
      <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panic.rs:308:9
      std::panicking::try::do_call
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:381:40
      std::panicking::try
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:345:19
      std::panic::catch_unwind
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panic.rs:382:14
      std::thread::Builder::spawn_unchecked::{{closure}}
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/thread/mod.rs:463:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/ops/function.rs:227:5
  20: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/boxed.rs:1042:9
      <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/boxed.rs:1042:9
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys/unix/thread.rs:87:17
  21: start_thread
  22: clone
    
materialized encountered an internal error and crashed.
```

File: test/testdrive/temporary.td
Comment (philip-stoev) [inline test/testdrive/temporary.td:121]: ```SHOW CREATE TABLE LIKE``` does not pick temporary tables.

```
pstoev=> create temporary table t1 (f1 integer);
CREATE TABLE

pstoev=> show tables like 't1';
 name 
------
(0 rows)

```

**Comment by philip-stoev** [comment]
@JLDLaughlin apologies for making a mess in the review, it would have looked more organized in Atlassian's review tools. I also managed to report multiple manifestations of the same underlying issues, so I think you will be able to close off the conversations in batches.

Please advise as to how would you like the tests to be delivered . I could push them to your fork directly (if granted access) either before or after your fixes, and then you can merge them upstream along with your code, but I am open to suggestions regarding workflow.

**Comment by JLDLaughlin** [comment]
@philip-stoev - thank you, I'm so glad you tested!

please feel free to push the tests to this branch whenever is best for you, I will work on fixes in the meantime.

File: test/testdrive/temporary.td
Comment (JLDLaughlin) [reply test/testdrive/temporary.td:70]: I didn't realize this!

> Existing permanent tables with the same name are not visible to the current session while the temporary table exists, unless they are referenced with schema-qualified names. Any indexes created on a temporary table are automatically temporary as well.

Since this is also true for views, this is also broken for our views:
```
materialize=> create view test as select 1;
CREATE VIEW
materialize=> create temp view test as select 2;
CREATE VIEW
materialize=> select * from test;
 ?column?
----------
        1
(1 row)
```

Comment (JLDLaughlin) [reply test/testdrive/temporary.td:70]: Fixed this locally for all temporary objects!

File: test/testdrive/temporary.td
Comment (JLDLaughlin) [reply test/testdrive/temporary.td:72]: Fixed the error message locally!

Comment (JLDLaughlin) [reply test/testdrive/temporary.td:72]: I think I'll punt the conversion to a temporary object to another PR, if you don't mind! I've created an issue here: https://github.com/MaterializeInc/materialize/issues/5580.

Comment (philip-stoev) [reply test/testdrive/temporary.td:72]: Yes totally please proceed as you see fit.

File: test/testdrive/temporary.td
Comment (JLDLaughlin) [reply test/testdrive/temporary.td:121]: I believe that all of our `SHOW` commands only show items in the current (or given schema) by default. For some `SHOW` commands, we have an `EXTENDED` option that will include items outside of the current schema, as well.

Do you think that implementing `SHOW EXTENDED TABLES` like this would be sufficient? Or should `SHOW TABLES` always include a connection's temporary tables?
```
pstoev=> create temporary table t1 (f1 integer);
CREATE TABLE

pstoev=> show extended tables like 't1';
 name 
------
(0 rows)
```

File: test/testdrive/temporary.td
Comment (JLDLaughlin) [reply test/testdrive/temporary.td:74]: I was able to fix this locally! (The same thing fixed this and the table masking issue you pointed out.)
```
materialize=> create table t1 (f1 integer);
CREATE TABLE
materialize=> create temporary table t1 (f1 integer);
CREATE TABLE
materialize=> drop table t1;
DROP TABLE
materialize=> explain select * from t1;
          Optimized Plan
----------------------------------
 %0 =                            +
 | Get materialize.public.t1 (u1)+

(1 row)
```

File: test/testdrive/temporary.td
Comment (JLDLaughlin) [reply test/testdrive/temporary.td:76]: Great catch! This is also an issue for views. For the curious, [the bug is here](https://github.com/MaterializeInc/materialize/blob/main/src/sql/src/ast/transform.rs#L37).

File: test/testdrive/temporary.td
Comment (philip-stoev) [reply test/testdrive/temporary.td:121]: EXTENDED is good enough, thank you.

Comment (JLDLaughlin) [reply test/testdrive/temporary.td:121]: Perfect! Created an issue for that here: https://github.com/MaterializeInc/materialize/issues/5593

File: test/testdrive/temporary.td
Comment (JLDLaughlin) [reply test/testdrive/temporary.td:76]: Fixed this locally! A more satisfying fix will come with https://github.com/MaterializeInc/materialize/issues/5591

File: test/testdrive/temporary.td
Comment (JLDLaughlin) [reply test/testdrive/temporary.td:71]: This issue is certainly related, but a bit removed from the code in this PR. I'm going to reopen our `\dt` issue and address it there: https://github.com/MaterializeInc/materialize/issues/4284

**Comment by philip-stoev** [comment]
I have pushed the tests, so it is all good. Thank you for your patience.

**Review by philip-stoev — APPROVED** [review]

**Comment by JLDLaughlin** [comment]
Sweet teamwork, thanks for your help @philip-stoev !

**Pull request merged**
